File: frontend/src/lib/api/inventory.ts
import { api, post, put, del, buildQuery } from './client';

export interface InventoryItem {
	id: number;
	name: string;
	category: string;
	unit: string;
	quantity: number;
	min_quantity: number;
	cost_per_unit: number | null;
	supplier: string | null;
	notes: string | null;
	created_at: string;
	updated_at: string;
}

export interface CreateInventoryItem {
	name: string;
	category: string;
	unit?: string;
	quantity?: number;
	min_quantity?: number;
	cost_per_unit?: number;
	supplier?: string;
	notes?: string;
}

export interface UpdateInventoryItem {
	name?: string;
	category?: string;
	unit?: string;
	min_quantity?: number;
	cost_per_unit?: number;
	supplier?: string;
	notes?: string;
}

export interface InventoryTransaction {
	id: number;
	item_id: number;
	transaction_type: string;
	quantity: number;
	notes: string | null;
	transaction_date: string;
	created_at: string;
}

export interface CreateInventoryTransaction {
	item_id: number;
	transaction_type: string;
	quantity: number;
	notes?: string;
	transaction_date?: string;
}

export interface InventoryFilter {
	category?: string;
	low_stock?: boolean;
	search?: string;
	page?: number;
	per_page?: number;
}

export const INVENTORY_CATEGORY_LABELS: Record<string, string> = {
	feed: 'Корма',
	medicine: 'Медикаменты',
	supplies: 'Расходники',
	equipment: 'Оборудование',
	other: 'Прочее',
};

export function listItems(filter: InventoryFilter = {}, signal?: AbortSignal) {
	return api<{ data: InventoryItem[]; total: number; page: number; per_page: number }>(
		`/inventory${buildQuery(filter)}`,
		{ signal },
	);
}

export function createItem(data: CreateInventoryItem) {
	return post<{ data: InventoryItem }>('/inventory', data);
}

export function updateItem(id: number, data: UpdateInventoryItem) {
	return put<{ data: InventoryItem }>(`/inventory/${id}`, data);
}

export function deleteItem(id: number) {
	return del<{ message: string }>(`/inventory/${id}`);
}

export function createTransaction(itemId: number, data: Omit<CreateInventoryTransaction, 'item_id'>) {
	return post<{ data: InventoryTransaction }>(`/inventory/${itemId}/transaction`, { ...data, item_id: itemId });
}

export function getLowStock(signal?: AbortSignal) {
	return api<{ data: InventoryItem[] }>('/inventory/low-stock', { signal });
}

File: backend/src/models/inventory.rs
use chrono::NaiveDate;
use serde::{Deserialize, Serialize};
use sqlx::FromRow;

use crate::errors::AppError;

#[derive(Debug, Clone, Serialize, Deserialize, FromRow, utoipa::ToSchema)]
pub struct InventoryItem {
    pub id: i32,
    pub name: String,
    pub category: String,
    pub unit: String,
    pub quantity: f64,
    pub min_quantity: f64,
    pub cost_per_unit: Option<f64>,
    pub supplier: Option<String>,
    pub notes: Option<String>,
    pub created_at: chrono::DateTime<chrono::Utc>,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
pub struct CreateInventoryItem {
    pub name: String,
    pub category: String,
    pub unit: Option<String>,
    pub quantity: Option<f64>,
    pub min_quantity: Option<f64>,
    pub cost_per_unit: Option<f64>,
    pub supplier: Option<String>,
    pub notes: Option<String>,
}

impl CreateInventoryItem {
    pub fn validate(&self) -> Result<(), AppError> {
        use crate::validation::*;
        required_non_empty(&self.name, "Название")?;
        max_len(&self.name, 200, "Название")?;
        required_non_empty(&self.category, "Категория")?;
        max_len(&self.category, 50, "Категория")?;
        opt_max_len(&self.unit, 20, "Единица")?;
        opt_non_negative_f64(&self.quantity, "Количество")?;
        opt_non_negative_f64(&self.min_quantity, "Мин. количество")?;
        opt_non_negative_f64(&self.cost_per_unit, "Цена за единицу")?;
        opt_max_len(&self.supplier, 200, "Поставщик")?;
        opt_max_len(&self.notes, 500, "Заметки")?;
        Ok(())
    }
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
pub struct UpdateInventoryItem {
    pub name: Option<String>,
    pub category: Option<String>,
    pub unit: Option<String>,
    pub min_quantity: Option<f64>,
    pub cost_per_unit: Option<f64>,
    pub supplier: Option<String>,
    pub notes: Option<String>,
}

impl UpdateInventoryItem {
    pub fn validate(&self) -> Result<(), AppError> {
        use crate::validation::*;
        if let Some(ref v) = self.name {
            required_non_empty(v, "Название")?;
            max_len(v, 200, "Название")?;
        }
        if let Some(ref v) = self.category {
            required_non_empty(v, "Категория")?;
            max_len(v, 50, "Категория")?;
        }
        opt_max_len(&self.unit, 20, "Единица")?;
        opt_non_negative_f64(&self.min_quantity, "Мин. количество")?;
        opt_non_negative_f64(&self.cost_per_unit, "Цена за единицу")?;
        opt_max_len(&self.supplier, 200, "Поставщик")?;
        opt_max_len(&self.notes, 500, "Заметки")?;
        Ok(())
    }
}

#[derive(Debug, Deserialize, utoipa::ToSchema, utoipa::IntoParams)]
pub struct InventoryFilter {
    pub category: Option<String>,
    pub low_stock: Option<bool>,
    pub search: Option<String>,
    pub page: Option<i64>,
    pub per_page: Option<i64>,
}

#[derive(Debug, Clone, Serialize, Deserialize, FromRow, utoipa::ToSchema)]
pub struct InventoryTransaction {
    pub id: i32,
    pub item_id: i32,
    pub transaction_type: String,
    pub quantity: f64,
    pub notes: Option<String>,
    pub transaction_date: NaiveDate,
    pub created_at: chrono::DateTime<chrono::Utc>,
}

#[derive(Debug, Deserialize, utoipa::ToSchema)]
pub struct CreateInventoryTransaction {
    pub item_id: i32,
    pub transaction_type: String,
    pub quantity: f64,
    pub notes: Option<String>,
    pub transaction_date: Option<NaiveDate>,
}

impl CreateInventoryTransaction {
    pub fn validate(&self) -> Result<(), AppError> {
        use crate::validation::*;
        positive_i32(self.item_id, "ID позиции")?;
        if self.quantity <= 0.0 {
            return Err(AppError::BadRequest(
                "Количество должно быть больше нуля".into(),
            ));
        }
        if self.transaction_type != "in"
            && self.transaction_type != "out"
            && self.transaction_type != "adjustment"
        {
            return Err(AppError::BadRequest(
                "Тип транзакции должен быть in, out или adjustment".into(),
            ));
        }
        opt_max_len(&self.notes, 500, "Заметки")?;
        Ok(())
    }
}
